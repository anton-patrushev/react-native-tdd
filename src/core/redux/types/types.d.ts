import { StateType, ActionType } from 'typesafe-actions';

// declare module 'MyTypes' {
//   export type Store = StateType<typeof import('./index').default>;
//   export type RootState = StateType<
//     ReturnType<typeof import('./root-reducer').default>
//   >;
//   export type RootAction = ActionType<typeof import('./root-action').default>;

//   export type RootEpic = Epic<RootAction, RootAction, RootState, Services>;
// }

declare module 'typesafe-actions' {
  interface Types {
    RootAction: ActionType<typeof import('../actions').default>;
    // Store: StateType<typeof store>;
    // RootState: StateType<ReturnType<typeof import('@/reducers').default>>;
  }
}
