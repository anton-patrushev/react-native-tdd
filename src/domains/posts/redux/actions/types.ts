export enum GetPostsActionTypes {
  REQUEST = 'posts/GET_POSTS_REQUEST',
  LOADING = 'posts/GET_POSTS_LOADING',
  FAILURE = 'posts/GET_POSTS_FAILURE',
  SUCCESS = 'posts/GET_POSTS_SUCCESS',
}
