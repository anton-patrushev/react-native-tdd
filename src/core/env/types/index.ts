export enum EnvironmentNames {
  DEVELOPMENT = 'development',
  STAGING = 'staging',
  QA = 'qa',
  PRODUCTION = 'production',
}

export type EnvironmentConfig = {
  ENVIRONMENT: EnvironmentNames;
  API_BASE: string;
  SOME_ANALYTICS_KEY: string;
};
