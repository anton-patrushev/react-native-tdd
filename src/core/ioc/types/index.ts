export enum Dependency {
  POSTS_REPOSITORY = 'postsRepository',
}
