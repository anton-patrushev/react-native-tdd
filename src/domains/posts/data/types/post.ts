export type Post = {
  id: number;
  // TODO: add userId relationship
  // userId: number;
  title: string;
  body: string;
};
