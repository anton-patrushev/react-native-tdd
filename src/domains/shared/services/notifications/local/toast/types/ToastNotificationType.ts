export enum ToastNotificationType {
  SUCCESS = 'success',
  ERROR = 'error',
  INFO = 'info',
}
