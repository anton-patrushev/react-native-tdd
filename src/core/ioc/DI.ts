import FakePostRepository from 'src/posts/api/network/FakePostsRepository';
import { IPostsRepository } from 'src/posts/api/network/IPostsRepository';

import { Dependency } from './types';

interface DIContainer {
  [Dependency.POSTS_REPOSITORY]: IPostsRepository;
}

const container: DIContainer = {
  [Dependency.POSTS_REPOSITORY]: FakePostRepository,
};

// Ignore from the coverage
/* istanbul ignore next */
function registerDependency<K extends keyof DIContainer>(
  key: K,
  dependency: DIContainer[K],
): void {
  container[key] = dependency;
}

// Ignore from the coverage
/* istanbul ignore next */
function getDependency<K extends keyof DIContainer>(key: K): DIContainer[K] {
  return container[key];
}

const DI = {
  registerDependency,
  getDependency,
};

export default DI;
