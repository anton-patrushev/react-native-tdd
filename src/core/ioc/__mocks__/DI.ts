import FakePostRepository from 'src/domains/posts/data/network/FakePostsRepository';

import FakeToastNotificationService from 'src/domains/shared/services/notifications/local/toast/FakeToastNotifcationService';

import { Dependency, DIContainer } from '../types';

const mockedContainer: DIContainer = {
  [Dependency.POSTS_REPOSITORY]: FakePostRepository,
  [Dependency.TOAST_NOTIFICATION_SERVICE]: FakeToastNotificationService,
};

const DI = {
  registerDependency: jest.fn(
    <K extends keyof DIContainer>(key: K, dependency: DIContainer[K]) => {
      mockedContainer[key] = dependency;
    },
  ),
  getDependency: jest.fn(
    <K extends keyof DIContainer>(key: K): DIContainer[K] => {
      return mockedContainer[key];
    },
  ),
};

export default DI;
