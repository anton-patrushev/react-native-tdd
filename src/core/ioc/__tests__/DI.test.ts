// yarn jest --coverage src/core/ioc/__tests__/DI.test.ts

import DI from '../DI';
import { Dependency } from '../types';

describe('DI', () => {
  it('should have initialValues before initialization', () => {
    const PostsRepository = DI.getDependency(Dependency.POSTS_REPOSITORY);
    expect(PostsRepository).toMatchSnapshot(Dependency.POSTS_REPOSITORY);

    const ToastNotificationService = DI.getDependency(
      Dependency.TOAST_NOTIFICATION_SERVICE,
    );
    expect(ToastNotificationService).toMatchSnapshot(
      Dependency.TOAST_NOTIFICATION_SERVICE,
    );
  });

  it('should register dependency successfully', () => {
    const dependencyKey = 'dependency-key';

    const createDependency = jest.fn(() => ({
      uniqueDependencyKey: dependencyKey,
      uniqueMethod: jest.fn(),
    }));

    // @ts-ignore (ignore type-check since no reason to register real dependency)
    DI.registerDependency(dependencyKey, createDependency());
    // @ts-ignore (ignore type-check since no reason to register real dependency)
    const dependency = DI.getDependency(dependencyKey) as ReturnType<
      typeof createDependency
    >;

    dependency.uniqueMethod();

    expect.assertions(2);
    expect(dependency.uniqueDependencyKey).toBe(dependencyKey);
    expect(dependency.uniqueMethod).toBeCalledTimes(1);
  });
});
