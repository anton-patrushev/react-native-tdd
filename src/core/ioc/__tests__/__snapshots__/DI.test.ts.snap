// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DI should have initialValues before initialization: postsRepository 1`] = `
FakePostsRepository {
  "getPosts": [Function],
}
`;

exports[`DI should have initialValues before initialization: toastNotificationService 1`] = `
_FakeNotificationService {
  "_showError": [Function],
  "_showInfo": [Function],
  "_showSuccess": [Function],
  "show": [Function],
}
`;
